File: src/test/java/cube8540/oauth/authentication/credentials/oauth/error/OAuth2ErrorSerializerTest.java
package cube8540.oauth.authentication.credentials.oauth.error;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.SerializerProvider;
import cube8540.oauth.authentication.credentials.oauth.OAuth2Utils;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.security.oauth2.core.OAuth2Error;
import org.springframework.security.oauth2.core.OAuth2ErrorCodes;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

class OAuth2ErrorSerializerTest {

    private static final String ERROR_CODE = OAuth2ErrorCodes.INVALID_REQUEST;

    private static final String ERROR_DESCRIPTION = "INVALID REQUEST";

    private OAuth2Error error;
    private JsonGenerator jsonGenerator;
    private SerializerProvider provider;

    private OAuth2ErrorSerializer serializer;

    @BeforeEach
    void setup() {
        this.error = new OAuth2Error(ERROR_CODE, ERROR_DESCRIPTION, null);
        this.jsonGenerator = mock(JsonGenerator.class);
        this.provider = mock(SerializerProvider.class);
        this.serializer = new OAuth2ErrorSerializer();
    }

    @Test
    @DisplayName("JsonGenerator 객체의 writeStartObject 메소드를 한번 호출해야 한다.")
    void shouldCallJsonGeneratorWriteStartObject() throws Exception {
        serializer.serialize(error, jsonGenerator, provider);

        verify(jsonGenerator, times(1)).writeStartObject();
    }

    @Test
    @DisplayName("JsonGenerator에 애러 코드를 적어야 한다.")
    void shouldWriteErrorCodeInJsonGenerator() throws Exception {
        serializer.serialize(error, jsonGenerator, provider);

        verify(jsonGenerator, times(1)).writeStringField(OAuth2Utils.ErrorMessageKey.ERROR, ERROR_CODE);
    }

    @Test
    @DisplayName("JsonGenerator에 에러 설명 적어야 한다.")
    void shouldWriteErrorDescriptionInJsonGenerator() throws Exception {
        serializer.serialize(error, jsonGenerator, provider);

        verify(jsonGenerator, times(1)).writeStringField(OAuth2Utils.ErrorMessageKey.DESCRIPTION, ERROR_DESCRIPTION);
    }

    @Test
    @DisplayName("JsonGenerator 객체의 writeEndObject 메소드를 호출해야 한다.")
    void shouldCallJsonGeneratorWriteEndObject() throws Exception {
        serializer.serialize(error, jsonGenerator, provider);

        verify(jsonGenerator, times(1)).writeEndObject();
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/oauth/error/OAuth2ErrorSerializerTest.kt
package cube8540.oauth.authentication.oauth.error

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.databind.SerializerProvider
import cube8540.oauth.authentication.oauth.ErrorMessageKey
import io.mockk.mockk
import io.mockk.verifyOrder
import org.junit.jupiter.api.Test
import org.springframework.security.oauth2.core.OAuth2Error
import org.springframework.security.oauth2.core.OAuth2ErrorCodes

class OAuth2ErrorSerializerTest {

    private val serializer = OAuth2ErrorSerializer()

    @Test
    fun `oauth2 error serialize`() {
        val error = OAuth2Error(OAuth2ErrorCodes.INVALID_REQUEST, "invalid request", null)
        val jsonGenerator: JsonGenerator = mockk(relaxed = true)
        val provider: SerializerProvider = mockk()

        serializer.serialize(error, jsonGenerator, provider)
        verifyOrder {
            jsonGenerator.writeStartObject()
            jsonGenerator.writeStringField(ErrorMessageKey.ERROR, OAuth2ErrorCodes.INVALID_REQUEST)
            jsonGenerator.writeStringField(ErrorMessageKey.DESCRIPTION, "invalid request")
            jsonGenerator.writeEndObject()
        }
    }

}